File: solutions/137.single-number-ii/single-number-ii.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int res = 0;
        for(int i = 0; i < 32; i++) {
            int mask = 1 << i;
            int count = 0;
            for(int num : nums) 
                if(num & mask) 
                    count++;
            if(count % 3) 
                res |= mask;
        }
        return res;
    }
};

File: solutions/137.single-number-ii/single-number-ii.java
class Solution {
    // public int singleNumber(int[] nums) {
    //     //eg: 01 the ones is 1, and the twos is 0;
    //     int ones = 0, twos = 0;
    //     for(int num : nums) {
    //         ones = (ones ^ num) & ~twos;
    //         twos = (twos ^ num) & ~ones;
    //     }
    //     return ones;
    // }
    public int singleNumber(int[] nums) {
        int ones = 0, twos = -1;
        for(int num : nums) {
            ones = ones ^ num & twos;
            twos = twos ^ num & ones;
        }
        return ones;
    }
}